src/app/ReadHandler.cpp

```diff
@@ -108,6 +103,11 @@ void ReadHandler::Close()
         mpExchangeCtx = nullptr;
     }
 
+    if (IsAwaitingReportResponse())
+    {
+        InteractionModelEngine::GetInstance()->GetReportingEngine().OnReportConfirm();
```

Comment:
I think this can get us in a bad state.  Consider `ReadHandler::OnStatusResponse` when we are in the `AwaitingReportResponse` state, `IsChunkedReport()` is false, we're a subscription, and `IsPriming()` is true.

We will call `OnReportConfirm` but remain in the `AwaitingReportResponse` state while we call `SendSubscribeResponse`.  If anything in there errors out before reaching the end and moving to the GeneratingReports state, the `exit` block in `OnStatusResponse` will call `Close`, which will call `OnReportConfirm` _again_.

Then we will end up underflowing the counter in InteractionModelEngine, and likely break reads completely.

We should ensure that we do not stay in the AwaitingReportResponse state after we call OnReportConfirm... preferably by having a single function both make that call and update the state.  In fact, can we do this in `MoveToState` as we move out of the AwaitingReportResponse state?

Reply:
agree, put it in moveToState, and add tests.